cmd/uninstall: check for dependent casks

nandahkrishna opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----

This PR aims to add the ability to check for casks dependent on formulae when running `brew uninstall`, as requested in #10570.

Output:
```
➜ brew uninstall mono                 
Error: Refusing to uninstall /usr/local/Cellar/mono/6.12.0.107
because it is required by godot-mono, which is currently installed.
You can override this and force removal with:
  brew uninstall --ignore-dependencies mono
```

~TODO: Comments for the new `find_cask_dependents` method and tests.~

MikeMcQuaid commented on Library/Homebrew/keg.rb:
```diff
@@ -190,6 +190,57 @@ def self.find_some_installed_dependents(kegs)
     [all_required_kegs.to_a, all_dependents.sort]
   end
 
+  # Given an array of kegs, this method finds casks dependent on them
+  # except those included in the `casks` array. If it does, it returns:
```
I think it'd be good to have this somewhere under `cask/` rather than in `keg` (which is already largish)

nandahkrishna replied:
Actually, it might be possible to merge this method with `Keg.find_some_installed_dependents` as some of the code is repetitive. Would that be okay too?

MikeMcQuaid replied:
I'm 👍🏻 on sharing code on them if it makes sense. I don't think there should be any references to casks in `keg.rb` though; casks never have/use kegs.

MikeMcQuaid commented on Library/Homebrew/uninstall.rb:
```diff
@@ -96,18 +97,27 @@ def uninstall_kegs(kegs_by_rack, force: false, ignore_dependencies: false, named
+      return false if result_kegs.blank? && result_casks.blank?
+
+      result = if result_kegs.present? && result_casks.present?
+        [(result_kegs[0] + result_casks[0]).uniq, result_kegs[1] + result_casks[1]]
```
The `[0]` and `[1]` here is a bit hard to follow here. Maybe use intermediate variables so it's more obvious what's going on here?

MikeMcQuaid reviewed (commented):
Great work so far!

nandahkrishna commented:
> I'm 👍🏻 on sharing code on them if it makes sense.

I've done this with the latest commit(s). I thought it would be better to have `Keg.find_some_installed_dependents` get dependents of kegs irrespective of whether those dependents are kegs themselves or casks. I hope my thinking isn't incorrect here.

> I don't think there should be any references to casks in keg.rb though; casks never have/use kegs.

If I am wrong and/or this point (**any** references) takes precedence, I'll move the handling to `cask/`.

MikeMcQuaid commented on Library/Homebrew/keg.rb:
```diff
@@ -137,18 +137,19 @@ def to_s
   #
   # For efficiency, we don't bother trying to get complete data.
-  def self.find_some_installed_dependents(kegs)
+  def self.find_some_installed_dependents(kegs, casks: [])
```
I'd rather avoid this method on `Keg` take `casks` because kegs and casks are completely separate concepts. I'd suggest this ensure function move to a new `installed_dependents.rb` or similar.

nandahkrishna commented:
With the latest commit, I've moved `find_some_installed_dependents` from `keg.rb` to a new `installed_dependents.rb` (hope I understood the suggestion right).

This method had quite a few tests and so I've moved them as-is from the previous `keg_spec.rb`, along with an additional test that is cask-related.

MikeMcQuaid reviewed (approved):
This is perfect, exactly what I had in mind. Great work @nandahkrishna 👏🏻 ❤️
